// @apollo/experimental-nextjs-app-support/dist/ssr/NextSSRApolloClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NextSSRApolloClient = void 0;
const client_1 = require("@apollo/client");
const graphql_1 = require("graphql");
const cache_1 = require("@apollo/client/cache");
const lateInitializingQueue_1 = require("./lateInitializingQueue");
const ApolloRehydrateSymbols_1 = require("./ApolloRehydrateSymbols");
function getQueryManager(client) {
    return client["queryManager"];
}
class NextSSRApolloClient extends client_1.ApolloClient {
    constructor(options) {
        super(options);
        this.rehydrationContext = {
            incomingBackgroundQueries: [],
        };
        this.resolveFakeQueries = new Map();
        this.registerWindowHook();
    }
    identifyUniqueQuery(options) {
        const transformedDocument = this.documentTransform.transformDocument(options.query);
        const queryManager = getQueryManager(this);
        // Calling `transformDocument` will add __typename but won't remove client
        // directives, so we need to get the `serverQuery`.
        const { serverQuery } = queryManager.getDocumentInfo(transformedDocument);
        if (!serverQuery) {
            throw new Error("could not identify unique query");
        }
        const canonicalVariables = (0, cache_1.canonicalStringify)(options.variables || {});
        const cacheKey = [(0, graphql_1.print)(serverQuery), canonicalVariables].toString();
        return { query: serverQuery, cacheKey, varJson: canonicalVariables };
    }
    registerWindowHook() {
        if (typeof window !== "undefined") {
            if (Array.isArray(window[ApolloRehydrateSymbols_1.ApolloBackgroundQueryTransport] || [])) {
                (0, lateInitializingQueue_1.registerLateInitializingQueue)(ApolloRehydrateSymbols_1.ApolloBackgroundQueryTransport, (options) => {
                    const { query, varJson, cacheKey } = this.identifyUniqueQuery(options);
                    if (!query)
                        return;
                    const printedServerQuery = (0, graphql_1.print)(query);
                    const queryManager = getQueryManager(this);
                    const byVariables = queryManager["inFlightLinkObservables"].get(printedServerQuery) ||
                        new Map();
                    queryManager["inFlightLinkObservables"].set(printedServerQuery, byVariables);
                    if (!byVariables.has(varJson)) {
                        const promise = new Promise((resolve, reject) => {
                            this.resolveFakeQueries.set(cacheKey, [resolve, reject]);
                        });
                        const cleanupCancelFn = () => queryManager["fetchCancelFns"].delete(cacheKey);
                        byVariables.set(varJson, new client_1.Observable((observer) => {
                            promise
                                .then((result) => {
                                observer.next(result);
                                observer.complete();
                            })
                                .catch((err) => {
                                observer.error(err);
                            })
                                .finally(() => {
                                this.resolveFakeQueries.delete(cacheKey);
                                cleanupCancelFn();
                            });
                        }));
                        queryManager["fetchCancelFns"].set(cacheKey, (reason) => {
                            var _a;
                            cleanupCancelFn();
                            const [_, reject] = (_a = this.resolveFakeQueries.get(cacheKey)) !== null && _a !== void 0 ? _a : [];
                            if (reject) {
                                reject(reason);
                            }
                        });
                    }
                });
            }
            if (Array.isArray(window[ApolloRehydrateSymbols_1.ApolloResultCache] || [])) {
                (0, lateInitializingQueue_1.registerLateInitializingQueue)(ApolloRehydrateSymbols_1.ApolloResultCache, (data) => {
                    var _a;
                    const { cacheKey } = this.identifyUniqueQuery(data);
                    const [resolve] = (_a = this.resolveFakeQueries.get(cacheKey)) !== null && _a !== void 0 ? _a : [];
                    if (resolve) {
                        resolve({
                            data: data.result,
                        });
                    }
                    // In order to avoid a scenario where the promise resolves without
                    // a query subscribing to the promise, we immediately call
                    // `cache.write` here.
                    // For more information, see: https://github.com/apollographql/apollo-client-nextjs/pull/38/files/388813a16e2ac5c62408923a1face9ae9417d92a#r1229870523
                    this.cache.write(data);
                });
            }
        }
    }
    watchQuery(options) {
        if (typeof window == "undefined") {
            if (options.fetchPolicy !== "cache-only" &&
                options.fetchPolicy !== "standby") {
                this.rehydrationContext.incomingBackgroundQueries.push(options);
            }
        }
        const result = super.watchQuery(options);
        return result;
    }
    setRehydrationContext(rehydrationContext) {
        if (rehydrationContext.incomingBackgroundQueries !==
            this.rehydrationContext.incomingBackgroundQueries)
            rehydrationContext.incomingBackgroundQueries.push(...this.rehydrationContext.incomingBackgroundQueries.splice(0));
        this.rehydrationContext = rehydrationContext;
    }
}
exports.NextSSRApolloClient = NextSSRApolloClient;
